rainbow-me/rainbow
# Multimedia NFT Support

This branch adds the capability to view or listen to rich media stored inside NFTs.

## 3D
  - The ability to detect whether an asset has a valid 3D `.glb` `animation_url`.
    - Unique assets which are detected in this format are represented alongside an additional icon in the `<AssetRecylerList />`.
    - The ability to render a 3D Model using `<model-viewer/>` using a `<WebView />`.
    - If the asset can be loaded successfully, it will be rendered.
    - If it possesses animation parameters, it will by default autoplay the animation.
    - Orbital controls are enabled to allow spherical rotation about the rendered scene.
    - Props are passed into the webpage using `JSON.stringify` to damage injection exploit vectors.
 
### Testing
1. Watch address `0x3b24804587a841c11b4e5baac8a6211671672ce0`
2. Navigate to **SuperRare collectibles**.
3. Tap on `Bernie Sanders` and watch him shiver. Zoom in closely to check out his `comfy gloves`.

## Video
  - If an asset possesses an `.mp4` `animation_url`, we’ll render the video content using `react-native-video`.
    - Does not `autoplay`.
    - Currently depends on a hack to manually stop the video when unmounted (likely the current native lib is getting confused when presenting within a `BottomSheet`).

### Testing
1. Watch address `0x188bd32f4c3b5ffaee22c2ec653f5d32ed27d844`
2. Navigate to **GeniusCorp collectibles**.
3. Watch the `Mighty33` video.

## Audio
  - If an asset possesses an `.mp3` `animation_url`, it’ll be detected as part of the user’s Playlist.
  - If the user has a non-empty Playlist, the `<EphemeralAudioPlayerFab />` will be displayed on the WalletScreen alongside the other FABs.
    - When no sound is currently playing, pressing play will load a random sound.
    - When a sound is playing, it can be paused, or the next track in the queue can be played by pressing fast forward.
    - When tracks end, it is configured to autoplay the next track in the queue.
    - Clicking on the title of the song will reopen the content in the `UniqueTokenExpandedView`.
  - Audio tokens also possess a Play/Pause button which is rendered within the `UniqueTokenExpandedView`. This can be used to configure runtime audio within the scope of the token itself.
    - Tokens are sensitive to their playback state; the Play/Pause button is expected to respond appropriately.
  - You can also listen to playing songs whilst the app is backgrounded. 
 
### Testing
1. Watch address `0x188bd32f4c3b5ffaee22c2ec653f5d32ed27d844`
2. Open up the sound FAB and hit play. Try cycling through tracks.
3. Click on a selected track's title.
4. Attempt to interact with the current playback.

## Added libraries
  - Added [use-debounce](https://www.npmjs.com/package/use-debounce) to help streamline the state transition between different pieces of audio content and prevent the user from downloading excessive audio content when cycling through their playlist.
  - Introduced [react-native-sound](https://github.com/zmxv/react-native-sound) to permit native level control of sound playback.
  - Introduced [react-native-video](https://github.com/react-native-video/react-native-video) to enable performant video downloading/caching.
